### DISCUSSION TITLE
Federation in a social network.

### DISCUSSION THREAD

**Discussion by ghost** [mastodon/mastodon#32701] [General]
Hi! :) Forgive me if this is the wrong forum for this sort of question. I have the following question: How exactly does federation work? I am familiar with how social networks work on non-federated basis. A user performs an action that is then saved in a DB and other users can see or interact with that action, depending on the rules of the platform. But assuming I am building a social network and would like to federate with other instances of the same social network, how would this work? Like on a technical level. I've done a few searches and tried a few things but they have all been dead ends.

**Comment by andypiper** [comment]
Mastodon uses the [ActivityPub specification](https://w3.org/TR/activitypub/) as part of its federation between server instances. It's a broad question though, and I would suggest that a site like [activitypub.rocks](https://activitypub.rocks/) would be better places to look for support.

**Comment by ghost** [comment]
@andypiper Hi Andy, thank you for pointing me towards those links and apologies for my late reply. Agreed, the way I posed the question was a bit too broad. What I am essentially trying to understand is this:

Let's say we have two servers running a software, let's call it FediSoft. Server A runs FediSoft and server B runs FediSoft too. Assuming the backend of FediSoft is an API service sitting on server A, that performs actions on the DB sitting on the same server based on the parameters supplied in the request, how does server B know about the activites happening on server A? For simplicity's sake, let's assume there are two tables in the DB handled by FediSoft's API service, a table for user accounts and a table for posts a user makes. How would this work? I roughly know what actors are and that they subscribe to an activity feed but what I fail to understand is how I should think of this subscription activity and the activity feed and how they relate to POST actions in the API service.

Forgive me if this asking a lot, but I'm really trying to understand this. I am building this in Rust, if that information is relevant.

**Reply by andypiper** [reply]
Again, I think this is not the right forum for your question.

If Server A knows about a user on Server B (because a user on Server A followed them), then messages will be published to Server A when things about the user on Server B are changed - new posts, deleted posts, etc.

For more on the internals of ActivityPub I'd suggest looking into [the new book by Evan Prodromou](https://www.ebooks.com/en-us/book/211467024/activitypub/evan-prodromou/). Chapter 4 covers server-to-server federation in detail.

**Comment by ghost** [comment]
Hi Andy, thank you. I'll see what I can rustle up from the pointers and information you have provided me with.
